### ISSUE TITLE
b-value scaling (again)

### ISSUE THREAD

**Issue by dchristiaens** [MRtrix3/mrtrix3#2231]
My apologies in advance to bring the b-value scaling up once again... 😉 

I came across a GE dataset where interspersed b-zero volumes are encoded like this:
```
$ mrinfo dwi.mif -dwgrad
...
   0.0475479964418854    -0.752325943629832    -0.657072950726055             2000.0003
    0.124123021599946     0.977383170399578     0.171218029799926           1999.999303
                    0                     0                     0                  2000
    0.603679064182912      0.28296303009199    -0.745321079178902           1999.999575
    0.422664756213416    -0.886664488528144     0.187564891805954           2000.002307
...
```
It appears that these volumes are not detected as b-zero scans. Since I can't be bothered reading through the lengthy discussions in the past threads on the subject, I just want to ask if there is any reason why this line should _not_ be interpreted as a b-zero volume?

The problem, I think, stems from these lines in the recent modifications:
https://github.com/MRtrix3/mrtrix3/blob/c4f5e0c995656cde8ea005c1a2a5105011c808e5/core/dwi/gradient.cpp#L259-L260
`squared_norms[row]` need to stay 0 to scale the b-value to 0 too. My suggestion would be to simply delete these lines. Any objections?

**Comment by jdtournier** [comment]
My one concern is the handling of Philips' isotropic-weighted volume that they sometimes include in the main series - [came up recently on the forum](https://community.mrtrix.org/t/dwi2response-error-iter0-fixel-peaks-mif/4010/4?u=jdtournier). I just had a look into our DICOM handling, and they're definitely included in the main series - there's nothing to really tell them apart from the other volumes in the series. But the only real option to deal with these is to remove them altogether, they simply shouldn't be included in the first place. With your changes, they'd be interpreted as b=0 volumes. Currently, they're stuff things up in other ways. Either way, they cause trouble, so that's actually not an argument against the change you're proposing...

So to cut a long story short: I think your changes are the only way to properly handle a single-shell scheme with interspersed b=0 volumes produced on systems that won't otherwise allow it (same as multi-shell handling, really). I'm guessing this change was added so that zero vectors wouldn't affect the automatic detection of whether b-value scaling should be applied, with the assumption that zero vectors won't come with a non-zero b-value? Either way, I reckon the correct fix here is to change that line 260 to:
```
squared_norms[row] = grad(row,3) > BZeroThreshold ? 0.0 : 1.0; 
```
In other words, if the b-value is non-zero, then b-value scaling should be applied (which leaving the norm as zero would trigger).

**Comment by Lestropie** [comment]
> My one concern is the handling of Philips' isotropic-weighted volume that they sometimes include in the main series

It seems to me that this relates to discussions regarding encoding of non-linear diffusion sensitisations. If these volumes were to be preserved, but the header encoding were to be faithful to the data, the gradient table would need to be projected up to the more general case (whatever form that ends up taking), with that final volume being explicitly classified as containing isotropic diffusion sensitisation (even though it's a _post hoc_ calculated trace-weighted image rather than being generated from isotropic diffusion sensitisation gradients).

That train of thought then exposes another consideration. Let's say that the diffusion gradient table is stored as complete *b*-tensors in this case, so that that final volume is encoded as isotropic. Now let's say that `mrconvert -coord 3 ` is used to remove that volume. Should the output image header still represent the gradient table using the full *b*-tensors even though only linear encodings remain, or should it instead go back to a 4-column `dw_scheme`?

(This reminds me of phase encoding handling, where the simpler BIDS-compatible header keys are used only if possible, with the more complex `pe_scheme` used only if necessary)

> But the only real option to deal with these is to remove them altogether, they simply shouldn't be included in the first place.

Maybe the better solution would be to, within the DICOM handler, artificially split such a volume into a separate DICOM series, such that the user must manually select whether they want all but the trace-weighted volume, or just the trace-weighted volume.

**Comment by dchristiaens** [comment]
Just to be clear, the image contrast in this case is in fact a regular non-diffusion-weighted (b=0) image, _not_ an isotropic-weighted volume. I noticed that you've made a separate issue for handling Philips isotropic scans in #2172.

As for this issue:

> I think your changes are the only way to properly handle a single-shell scheme with interspersed b=0 volumes produced on systems that won't otherwise allow it (same as multi-shell handling, really).

Exactly, I think it is the right thing to do. Isotropic-weighted data will always need to be marked in some other way.

> I'm guessing this change was added so that zero vectors wouldn't affect the automatic detection of whether b-value scaling should be applied, with the assumption that zero vectors won't come with a non-zero b-value? Either way, I reckon the correct fix here is to change that line 260 to:
> 
> ```
> squared_norms[row] = grad(row,3) > BZeroThreshold ? 0.0 : 1.0; 
> ```
> 
> In other words, if the b-value is non-zero, then b-value scaling should be applied (which leaving the norm as zero would trigger).

A check on the BZeroThreshold can't hurt, but it is not needed. I've tested 7070c6b on data with `0 0 0 0` encoding and everything works perfectly fine.

**Comment by dchristiaens** [comment]
Resolved in #2329
